File: src/main/kotlin/xyz/jmullin/drifter/extensions/RandomUtil.kt
package xyz.jmullin.drifter.extensions

import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.math.Vector3
import java.util.*

/**
 * Convenience methods for getting simple randomized results. Random functions require an implicit
 * [[Random]] instance. This allows for a fixed seed if repeatable random behavior is desirable.
 * A global [[Random]] instance with a nanotime seed is available at RandomUtil.Implicits.global if you prefer
 * not to supply your own.
 */
var r = Random()

fun <T> p(p: Number, a: T): T? = if (probability(p.toDouble())) a else null
fun <T> p(p: Number, a: () -> T): T? = if (probability(p.toDouble())) a() else null
fun <T> p(p: Number, a: () -> T, b: () -> T): T = if (probability(p.toDouble())) a() else b()
fun <T> p(p: Number, a: T, b: () -> T): T = if (probability(p.toDouble())) a else b()
fun <T> p(p: Number, a: () -> T, b: T): T = if (probability(p.toDouble())) a() else b
fun <T> p(p: Number, a: T, b: T): T = if (probability(p.toDouble())) a else b
fun probability(p: Double) = r.nextDouble() <= p
fun probability(p: Float) = r.nextFloat() <= p
fun rInt(n: Int) = r.nextInt(n)
fun rInt(n:Int, m:Int) = if(n == m) n else n+r.nextInt(m-n)
fun rInt(r: ClosedRange<Int>) = rInt(r.start, r.endInclusive)
fun rFloat(n: Float) = r.nextFloat()*n
fun rFloat(n: Float, m: Float): Float = n+r.nextFloat()*(m-n)
fun rFloat(r: ClosedRange<Float>) = rFloat(r.start, r.endInclusive)

fun <T> rElement(s: Iterable<T>) = s.toList().let { it[rInt(it.size)] }
fun <T> rElement(vararg s: T) = s.toList().let { it[rInt(it.size)] }
fun <T> Iterable<T>.shuffle(): List<T> {
    val list = this.toMutableList()
    Collections.shuffle(list, r)
    return list
}

fun rV(v: Vector2) = V2(rFloat(v.x), rFloat(v.y))
fun rV(a: Vector2, b: Vector2) = V2(rFloat(a.x, b.x), rFloat(a.y, b.y))
fun rV(r: Rectangle): Vector2 = rV(r.position, r.size)
fun rV(v: Vector3) = V3(rFloat(v.x), rFloat(v.y), rFloat(v.z))
fun rV(a: Vector3, b: Vector3) = V3(rFloat(a.x, b.x), rFloat(a.y, b.y), rFloat(a.z, b.z))
fun rColor(n: Float, m: Float) = C(rFloat(n, m), rFloat(n, m), rFloat(n, m))

fun ClosedRange<Float>.random() = rFloat(start, endInclusive)
fun ClosedRange<Int>.random() = rInt(start, endInclusive)

File: src/main/scala/xyz/jmullin/drifter/RandomUtil.scala
package xyz.jmullin.drifter

import xyz.jmullin.drifter.extensions.RichColor._
import xyz.jmullin.drifter.extensions.RichGeometry._

import scala.util.Random

/**
 * Convenience methods for getting simple randomized results. Random functions require an implicit
 * [[scala.util.Random]] instance. This allows for a fixed seed if repeatable random behavior is desirable.
 * A global [[Random]] instance with a nanotime seed is available at RandomUtil.Implicits.global if you prefer
 * not to supply your own.
 */
object RandomUtil {
  object Implicits {
    val global = new Random()
  }

  fun probability(p: Double)(implicit r: Random) = r.nextDouble() <= p
  fun rInt(n: Int)(implicit r: Random) = r.nextInt(n)
  fun rInt(n:Int, m:Int)(implicit r: Random) = if(n == m) n else n+r.nextInt(m-n)
  fun rFloat(n: Float)(implicit r: Random) = r.nextFloat()*n
  fun rFloat(n: Float, m: Float)(implicit r: Random) = n+r.nextFloat()*(m-n)
  fun rElement[T](s: Iterable[T])(implicit r: Random) = r.shuffle(s).head

  fun rV(v: Vector2)(implicit r: Random) = V2(rFloat(v.x)(r), rFloat(v.y)(r))
  fun rV(a: Vector2, b: Vector2)(implicit r: Random) = V2(rFloat(a.x, b.x)(r), rFloat(a.y, b.y)(r))
  fun rV(r: Rect)(implicit ra: Random): Vector2 = rV(r.v, r.size)(ra)
  fun rV(v: V3)(implicit r: Random) = V3(rFloat(v.x)(r), rFloat(v.y)(r), rFloat(v.z)(r))
  fun rV(a: V3, b: V3)(implicit r: Random) = V3(rFloat(a.x, b.x)(r), rFloat(a.y, b.y)(r), rFloat(a.z, b.z)(r))
  fun rColor(n: Float, m: Float)(implicit r: Random) = C(rFloat(n, m)(r), rFloat(n, m)(r), rFloat(n, m)(r))

  fun shuffle[T](seq: Seq[T])(implicit r: Random) = r.shuffle(seq)
}
